### PR TITLE
Refactor/store as hash

### PR DESCRIPTION
**Pull request by choonkeat**
- Core library code moved to https://github.com/choonkeat/attache_api (code here is specific to ActiveRecord)
- Add support for `:json` column
  - Changed to work with `Hash` and `[Hash, Hash]` instead of `String` and `[String, String]`

### CODE REVIEW

**Comment by choonkeat** [comment]
v2 data can be upgraded via migration generator

``` bash
rails g attache_rails:upgrade_v2_to_v3
```

which basically pulls out the values, change the `String` into `Hash` via `JSON.parse(String)` and save it back into the db

bulk of the work of the generator is to figure out which tables & columns are attache columns

**Comment by choonkeat** [comment]
@shinnyx @zamakkat fyi, though i know u're using a local `vendor/gems` copy. part of the reason for the breaking change is https://github.com/choonkeat/attache_rails/blob/refactor/store_as_hash/README.md#postgresql

**Comment by shinnc** [comment]
:+1: on the `json` column. Just wondering, why don't use `jsonb` which can be used for indexing?

Just learnt that it only supported on rails `4.2+` with extension.

**Comment by choonkeat** [comment]
> jsonb

i could add a note in README; support should be out of the box

File: README.md [hunk 1]
```diff
@@ -135,6 +137,35 @@ or
...
+- in previous version, `@user.photos` value could be something like `["{\"path\":\"dirname/file.jpg\"}"]`. Notice that it is an array of 1 `String`.
+- in v3, the value would be `[{"path"=>"dirname/file.jpg"}]`. Notice that it is an array of 1 `Hash`
+
+If you're upgrading from V2, we have a generator that will create a migration file to fixup the data
+
+```
+rails g attache_rails:upgrade_v2_to_v3
+```
+
+NOTE: It is highly recommended that developers verify the migration with a dump of the production data in a staging[] environment. Please take a look at the generated migration file.
```
Comment (winston) [inline README.md]: Is the `staging[]` a typo? i.e. just `staging`?

**Comment by winston** [comment]
Looks good to me!

File: lib/attache_rails/model.rb [hunk 2]
```diff
@@ -1,150 +1,52 @@
...
-        serialize name, JSON
-        define_method "#{name}_options",    -> (geometry, options = {}) { Utils.attache_options(geometry, Utils.array(self.send("#{name}_attributes", geometry)), multiple: false, **options) }
-        define_method "#{name}_url",        -> (geometry) {               self.send("#{name}_attributes", geometry).try(:[], 'url') }
-        define_method "#{name}_attributes", -> (geometry) {               str = self.send(name); Utils.attache_url_for(str, geometry) if str; }
-        define_method "#{name}=",           -> (value)    {
-          new_value = (value.respond_to?(:read) ? Utils.attache_upload_and_get_json(value) : value)
-          okay = JSON.parse(new_value)['path'] rescue nil
-          super(Utils.array(okay ? new_value : nil).first)
-        }
-        define_method "#{name}_discard_was",-> do
```
Comment (winston) [inline lib/attache_rails/model.rb:101]: We don't need the `*_was` methods anymore?

Comment (winston) [inline lib/attache_rails/model.rb:101]: Guess it's already handled by `after_update` right?

File: README.md (hunk 1 above)
Comment (choonkeat) [inline README.md]: yes typo. thanks for catching

File: lib/attache_rails/model.rb (hunk 2 above)
Comment (choonkeat) [inline lib/attache_rails/model.rb:101]: yes. we're doing "mark and sweep"
- during `after_update` we collect obsolete values from `old_values - new_values`
- during `after_destroy` we collect all current values as obsolete

then finally, everything is "swept" at `after_commit`

**Pull request merged**